src/Kestrel.Core/Internal/Http2/Http2FrameReader.cs

```diff
@@ -3,40 +3,245 @@
 
 using System;
 using System.Buffers;
+using System.Buffers.Binary;
+using System.Collections.Generic;
+using System.Diagnostics;
 
 namespace Microsoft.AspNetCore.Server.Kestrel.Core.Internal.Http2
 {
-    public static class Http2FrameReader
+    public class Http2FrameReader
     {
-        public static bool ReadFrame(ReadOnlySequence<byte> readableBuffer, Http2Frame frame, uint maxFrameSize, out SequencePosition consumed, out SequencePosition examined)
+        /* https://tools.ietf.org/html/rfc7540#section-4.1
+            +-----------------------------------------------+
+            |                 Length (24)                   |
+            +---------------+---------------+---------------+
+            |   Type (8)    |   Flags (8)   |
+            +-+-------------+---------------+-------------------------------+
+            |R|                 Stream Identifier (31)                      |
+            +=+=============================================================+
+            |                   Frame Payload (0...)                      ...
+            +---------------------------------------------------------------+
+        */
+        public const int HeaderLength = 9;
+
+        private const int LengthOffset = 0;
+        private const int TypeOffset = 3;
+        private const int FlagsOffset = 4;
+        private const int StreamIdOffset = 5;
+
+        public const int SettingSize = 6; // 2 bytes for the id, 4 bytes for the value.
+
+        // No frame type needs more than 8 bytes of additional fields. Payload data beyond the known fields is not stored in this buffer.
+        private readonly byte[] _extendedHeader = new byte[HeaderLength + 8];
+
+        public bool ReadFrame(ReadOnlySequence<byte> readableBuffer, Http2Frame frame, uint maxFrameSize, out ReadOnlySequence<byte> framePayload)
         {
-            consumed = readableBuffer.Start;
-            examined = readableBuffer.End;
+            framePayload = ReadOnlySequence<byte>.Empty;
 
-            if (readableBuffer.Length < Http2Frame.HeaderLength)
+            if (readableBuffer.Length < HeaderLength)
             {
                 return false;
             }
 
-            var headerSlice = readableBuffer.Slice(0, Http2Frame.HeaderLength);
-            headerSlice.CopyTo(frame.Raw);
+            var headerSlice = readableBuffer.Slice(0, HeaderLength);
+            headerSlice.CopyTo(_extendedHeader);
```

Comment:
Why do you copy at all?

Reply:
Because parsing out of a `ReadOnlySequence<byte>` looked more complex than these 15 bytes were worth.

Reply:
You don't need to copy though, you can just do what we do elsewhere.  Use the [ToSpan](https://github.com/aspnet/KestrelHttpServer/blob/d3d7c5519810ad97997f777c8a3683a9adbbe1e1/src/Kestrel.Core/Internal/Http/PipelineExtensions.cs#L20) extension method and 80% of the time it won't allocate. 😄

Reply:
That's a question of odds then. One allocation per connection, or one allocation every time it ToSpan fails. If it fails more than once per connection then you're net negative.

Also we have to do it in two segments since the extended headers are variable length.

Reply:
Does this logic run once per connection?

Reply:
No, it's on every frame.

Reply:
Then it's worth changing.